pkg/cli/mirror/util.go

```diff
@@ -113,3 +128,326 @@ func (o *MirrorOptions) checkErr(err error, acceptableErr func(error) bool, logM
 
 	return nil
 }
+
+func getPlatformKey(typeWithPlatorm interface{}) (*OperatorCatalogPlatform, error) {
+	if typeWithPlatorm == nil {
+		return nil, errors.New("no manifest or config provided, so unable to generate a platform key")
+	}
+	switch v := typeWithPlatorm.(type) {
+	case *v1.Platform:
+		return &OperatorCatalogPlatform{
+			os:           v.OS,
+			architecture: v.Architecture,
+			variant:      v.Variant,
+			isIndex:      true,
+		}, nil
+	case *v1.Descriptor:
+		if v.Platform == nil {
+			return nil, errors.New("no platform provided in descriptor, so unable to generate a platform key")
+		}
+		return &OperatorCatalogPlatform{
+			os:           v.Platform.OS,
+			architecture: v.Platform.Architecture,
+			variant:      v.Platform.Variant,
+			isIndex:      true,
+		}, nil
+	case *v1.ConfigFile:
+		// ConfigFile only comes into play when an image is a single architecture catalog.
+		// Callers may need to override the isIndex field if necessary
+		return &OperatorCatalogPlatform{
+			os:           v.OS,
+			architecture: v.Architecture,
+			variant:      v.Variant,
+			isIndex:      false,
+		}, nil
+	default:
+		// should not happen... this is a programmer error
+		return nil, fmt.Errorf("expected a manifest or config, but %T provided, so unable to generate a platform key", v)
+	}
+}
+
+/*
+getImageDigests will fetch one or more image digests for a given image reference.
+This function handles both "manifest list" and "image" references. A "manifest list"
+can return one or more digest references, whereas a "image" reference only returns a single
+digest reference.
+
+# Arguments
+
+• ctx: A cancellation context
+
+• imageRef: a docker image reference which will be either remotely accessed (if necessary)
+and used as a basis for the image references returned from this function
+
+• layoutPath: an optional OCI layout path. If provided imageRef will NOT be fetched remotely.
+This OCI layout path could refer to a "manifest list" or an "image".
+
+• insecure: flag to indicate if plain http is used or when skipping TLS verification
+
+# Returns
+
+• map[OperatorCatalogPlatform][]name.Digest: If no error occurs, returns a map whose key is Platform
+and value is a slice of one or more image digest references for that architecture.
+If an error occurs the map will always be initialized (i.e non-nil) but could have partial results.
+
+• error: non-nil if an error occurs, nil otherwise
+*/
+func getImageDigests(ctx context.Context, imageRef string, layoutPath *layout.Path, insecure bool) (map[OperatorCatalogPlatform][]name.Reference, error) {
+
+	// initialize return values
+	digestsMap := map[OperatorCatalogPlatform][]name.Reference{}
```

Comment:
Could there be more than one manifest per platform?
[]name.Reference contains the reference to manifest for that specific platform, not all digests. There's something I'm missing

Reply:
I did it this way to be more defensive about the references we discover along the way. There's two scenarios:

1. somebody duplicates a manifest entry for a given platform
2. nested "image indexes" are discovered resulting in a duplicate digest for a given platform

Neither of these are very likely, and I can totally see the confusion this causes. I can change this to a single reference rather than a slice if you'd prefer.

Reply:
I've refactored things, and as a result this is now a single reference rather than a slice.